### PR TITLE
periodic bc

### PR DESCRIPTION
**Pull request by tangqi**
Add two boundary conditions, for the cases I tested (advection and Poisson), PeriodicBC3 works the best.

Let me know if you catch any bugs, or you want me to make some changes on format. Thanks.

### CODE REVIEW

**Comment by lululxvi** [comment]
Some suggestions:

1. Remove PeriodicBC2;
2. Merge PeriodicBC3 into the current PeriodicBC, and add a new argument derivative_order to indicate the order of derivative, i.e.,
```
def __init__(self, geom, component_x, on_boundary, component=0, derivative_order=0):
```
3. Python file should be formatted by [black](https://github.com/psf/black)

**Comment by tangqi** [comment]
OK, I am fixing it based on your suggestion. I am testing it in a few cases, but I found an issue.

Here is your implementation:
```python
    def error(self, X, inputs, outputs, beg, end):
        outputs = outputs[beg:end, self.component : self.component + 1]
        outputs = tf.reshape(outputs, [-1, 2])
        return outputs[:, 0:1] - outputs[:, 1:]
```
Here is what I implement (no derivate)
```python
    def error(self, X, inputs, outputs, beg, end):
        mid=beg+int((end-beg)/2)
        yleft = outputs[beg:mid, self.component : self.component + 1]
        yright = outputs[mid:end, self.component : self.component + 1]
        return yleft-yright
```

In 1d they are equivalent. But in 2d, they are not. Maybe I miss something, but I thought in your implementation, the left and right solutions are stack on top of each other through `vstack`. Is the first one a bug? Or did I misunderstand something?

**Comment by tangqi** [comment]
BTW, for the periodic Poisson case, the second implementation gives me the correct answer, but the first one gives a wrong answer.

File: deepxde/boundary_conditions.py [hunk 1]
```diff
@@ -87,19 +87,32 @@ class PeriodicBC(BC):
...
     """
 
-    def __init__(self, geom, component_x, on_boundary, component=0):
+    def __init__(self, geom, component_x, on_boundary, derivative_order=0, component=0):
         super(PeriodicBC, self).__init__(geom, on_boundary, component)
         self.component_x = component_x
+        self.derivative_order = derivative_order
+        if derivative_order > 1:
+            raise NotImplementedError(
+                "PeriodicBC only support derivative_order 0 or 1."
```
Comment (lululxvi) [inline deepxde/boundary_conditions.py:96]: supports

**Comment by lululxvi** [comment]
Nice. Your implementation is correct. See more comments.

**Comment by tangqi** [comment]
Thanks for the suggestion. Fixed.

**Pull request merged**